opennlp-tools/src/main/java/opennlp/tools/util/DownloadUtil.java

```diff
@@ -80,6 +81,41 @@ public enum ModelType {
 
   private static Map<String, Map<ModelType, String>> availableModels;
 
+  /**
+   * Checks if a model of the specified {@code modelType} has been downloaded already
+   * for a particular {@code language}.
+   *
+   * @param language  The ISO language code of the requested model.
+   * @param modelType The {@link DownloadUtil.ModelType type} of model.
+   * @return {@code true} if a model exists locally, {@code false} otherwise.
+   * @throws IOException Thrown if IO errors occurred or the model is invalid.
+   */
+  static boolean existsModel(String language, ModelType modelType) throws IOException {
+    Map<ModelType, String> modelsByLanguage = getAvailableModels().get(language);
+    if (modelsByLanguage == null) {
+      return false;
+    } else {
+      final String url = modelsByLanguage.get(modelType);
+      if (url != null) {
+        final Path homeDirectory = Paths.get(System.getProperty("OPENNLP_DOWNLOAD_HOME",
+                System.getProperty("user.home"))).resolve(".opennlp");
+        final String filename = url.substring(url.lastIndexOf("/") + 1);
+        final Path localFile = Paths.get(homeDirectory.toString(), filename);
+        boolean exists;
+        try {
```

Comment:
```suggestion
 if (Files.exists(localFile)) {
          // if this does not throw the requested model exists AND is valid!
          validateModel(new URL(url + ".sha512"), localFile);
          return true;
        } else {
          return false;
        }
```

Would remove the "control flow by exception" here and do an additional check via `Files` since we already have the full path available. If it does not exist, we can fully skip the (hidden) download of a sha512 instead of failing after the actual sha512 download has happened.

Reply:
Side note: A future enhancement could be to actually download the sha512 to disk and store it alongside the model files, so we can avoid re-fetching the sha512 file all the time.